core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,255 @@
+                    upstream.cancel();
+                    return;
+                }
+            }
+        }
+
+        private void sendTail() {
+            setState(State.SENDING_TAIL, State.SENDING_COMPLETE);
+            if (tail != null) {
+                requestedUpdater.decrementAndGet(this);
+                downstream.onNext(tail);
+            } else {
+                sendComplete();
+            }
+        }
+
+        private void sendComplete() {
+            if (state == State.DONE) {
+                return;
+            }
+            setState(State.SENDING_COMPLETE, State.DONE);
+            downstream.onComplete();
+        }
+
+        private void requestUpstream(Subscription subscription) {
+            for (;;) {
+                final long requested = this.requested;
+                if (requested == 0) {
+                    return;
+                }
+                if (state == State.SENDING_TAIL) {
+                    sendTail();
+                    if (requested > 1) {
+                        sendComplete();
+                    }
+                    return;
+                }
+                if (requestedUpdater.compareAndSet(this, requested, 0)) {
+                    if (needToPublishUpdater.get(this) > Long.MAX_VALUE - requested) {
+                        needToPublishUpdater.set(this, Long.MAX_VALUE);
+                    } else {
+                        needToPublishUpdater.addAndGet(this, requested);
+                    }
+                    subscription.request(requested);
+                    return;
+                }
+            }
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (cancelled) {
+                subscription.cancel();
+                return;
+            }
+            upstream = subscription;
+            requestUpstream(subscription);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            needToPublishUpdater.decrementAndGet(this);
+            downstream.onNext(item);
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+            if (cancelled) {
+                return;
+            }
+            cancelled = true;
+            downstream.onError(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            final long needToPublish = needToPublishUpdater.get(this);
+            setState(State.SENDING_BODY, State.SENDING_TAIL);
+            if (needToPublish == 0) {
+                return;
+            }
+
+            sendTail();
+            if (needToPublish > 1) {
+                sendComplete();
+            }
+        }
+
+        @Override
+        public void cancel() {
+            if (cancelled) {
+                return;
+            }
+            cancelled = true;
```

Comment:
`head` and `tail` should be closed if they are not published to the downstream.

Reply:
You mean "we should send `head`&`tail` to downstream when they're not published" when `cancelled`?

I found that some publisher TCK tests failed on this way 🤔 (request -> cancel -> expectNoEvent)

Reply:
✅ updated!